docs/proposals/status-aggregation.md

```diff
@@ -10,39 +10,127 @@ When multiple instances of a resource (e.g. a Gateway) are running across multip
+In addition, if the `Ready` status condition type is `False`, the `Gateway` in the control plane should include a status message for each `Gateway` instance where `Ready` is `False`. This message would indicate the reason why the condition is not true for each `Gateway`.
 
 ```yaml
 status:
   conditions:
-  - type: ListenersValid
+  - type: Ready
     status: False
     message: "gateway-1 Listener certificate is expired; gateway-3 No listener configured for port 80"
 ```
 
-In this example, the `ListenersValid` status condition type is `False` because two of the three Gateway instances (gateway-1 and gateway-3) have issues with their listeners. For gateway-1, the reason for the `False` condition is that the listener certificate is expired, and for gateway-3, the reason is that no listener is configured for port 80. These reasons are included as status messages in the `Gateway` resource in the control plane.
+In this example, the `Ready` status condition type is `False` because two of the three Gateway instances (gateway-1 and gateway-3) have issues with their listeners. For gateway-1, the reason for the `False` condition is that the listener certificate is expired, and for gateway-3, the reason is that no listener is configured for port 80. These reasons are included as status messages in the `Gateway` resource in the control plane.
 
 As there may be different reasons for the condition being `False` across different clusters, it doesn't make sense to aggregate the `reason` field. The `reason` field is intended to be a programatic identifier, while the `message` field allows for a human readable message i.e. a semi-colon separated list of messages.
 
-The `lastTransitionTime` and `observedGeneration` fields will behave as normal for the resource in the control plane.
\ No newline at end of file
+The `lastTransitionTime` and `observedGeneration` fields will behave as normal for the resource in the control plane.
+
+### Addresses and Listeners status
+
+The Gateway status can include information about addresses, like load balancer IP Addresses assigned to the Gateway,
+and listeners, such as the number of attached routes for each listener.
+This information is useful at the control plane level.
+For example, a DNS Record should only exist as long as there is at least 1 attached route for a listener.
+It can also be more complicated than that when it comes to multi cluster gateways.
+A DNS Record should only include the IP Addresses of the Gateway instances where the listener has at least 1 attached route.
+This is important when initial setup of DNS Records happen as applications start.
+It doesn't make sense to route traffic to a Gateway where a listener isn't ready/attached yet.
+It also comes into play when a Gateway is displaced either due to changing placement decision or removal.
+
+In summary, the IP Addresses and number of attached routes per listener per Gateway instance is needed in the control plane to manage DNS effectively.
```

Comment:
would it make sense to say, an IP can only be added to the control plane if there is at least 1 listener on that gateway? That way any IP in the control plane can be used for a DNSRecord for any listener that has > 1 route 

So perhaps we can keep the structure the same at the control plane level?


```
status:
  addresses:
  - type: IPAddress
    value: 172.32.200.0 #ip from gateway instance with attached route for web
  - type: IPAddress
    value: 172.32.201.0 #ip from gateway instance with attached route for web
  listeners:
  - attachedRoutes: 0
    conditions:
    name: api
  - attachedRoutes: 2
    conditions:
    name: web
```

Reply:
Would this allow for the following scenario?

2 apps, (so 2 listeners) a.example.com & b.example.com were configured on 2 different clusters, Cluster X & Cluster Y.
The listener for app a is attached in Cluster X, but not Cluster Y.
The listener for app b is attached in both clusters.
If the IP Address for Cluster Y is added to the dns record for app a, some traffic will get a 503.

Reply:
So perhaps we can add our own condition type that way we don't need to actually modify the CRD?
```
lastTransitionTime: "2023-06-22T07:02:05Z"
message: "cluster1 listener api ready to recieve traffic"
observedGeneration: 1
reason: "listener:api,cluster:c1,addresses: 192.168.22.1,aws.lb.com"
status: "True"
type: MCAddress
```

Not ideal either but at least with this we avoid a change to the type? WDYT @mikenairn @david-martin

Reply:
@maleck13 presumably you can have many of these MCAddress type status conditions, one for each listener on each cluster?

Reply:
> So perhaps we can add our own condition type that way we don't need to actually modify the CRD?

Modifying the CRD isn't ideal.
Without modifying, we're bound by [the spec](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.GatewayStatus).

There's 3 fields we can work with: `addresses`, `conditions` and `listeners`.
Your suggestion to use conditions looks interesting.

> presumably you can have many of these MCAddress type status conditions

I haven't tried to use more than 1 condition with the same `type`, and it isn't explicitily forbidden [in the spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#condition-v1-meta). 
It kind of feels against what the intention of condition types are.

I wonder if we encoded the listener & cluster into the type so they are unique.
Something like:

```yaml
reason: "192.168.22.1"
status: "True"
type: c1.api/MCAddress
```

and for attachedRoutes:

```yaml
reason: "3"
status: "True"
type: c1.api/AttachedRoutes
```


Alternatively, we could do similar naming in the addresses and listeners fields (instead of conditions) to store the state of all gateways.

e.g.

```yaml
status:
  addresses:
  - type: c1.api/IPAddress
    value: 192.168.26.1
  - type: c1.api/IPAddress
    value: 192.168.26.2
  - type: c2.api/IPAddress
    value: 192.168.27.1
  listeners:
  - name: c1.api
    attachedRoutes: 2
  - name: c2.api
    attachedRoutes: 0
```

I'm trying to think if there's any reason these approaches wouldn't work...

Reply:
the types in the address status have to be an `AddressType` perhaps the value could carry the format?

Reply:
Adding a prefix of  cluster name to the listener, and prefixing the cluster and listener name to the value of the address would work in allowing us to get all the required information for setting up DNS. Seems a better option than trying to work it out from conditions, or creating a new CRD IMO.

Reply:
Maybe we want to use a `NamedAdress` type here https://github.com/robscott/gateway-api/blob/54a3bc3a01acae09330db5dddc179d335e36961a/apis/v1alpha2/gateway_types.go#L530? If I'm understanding what that is saying, it seems kind of close to what we want, a custom way of describing an address.

Reply:
yeah except in the comments that is marked as depreciated now

Reply:
either way lets go for a custom address @david-martin

Reply:
reading it  again we should be able to define our own addressType so this would work as you suggested @david-martin.

Reply:
I see, i saw that but couldn't work out what it was saying. It looks like we can create our own custom type as long as it passes this validation https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/shared_types.go#L538. So i guess we could have a `kuadrant.io/IPAddress` and `kuadrant.io/Hostname` and the value of these types has the values with the prefixed meta required to work out what cluster and listener it is.

Reply:
> with the prefixed meta

so something like, keeping the type static and the value with all the gateway specific data?:

```yaml
  - type: kuadrant.io/MultiClusterIPAddress
    value: gateway_name/192.168.26.1
```

Reply:
Yeah i think it is just cluster name we need here, and a listener for each cluster i.e.
```
  listeners:
  - name: cluster_name.api
    attachedRoutes: 2
  - name: cluster_name.api
    attachedRoutes: 0
```